File: front/lib/domain/sync/mutation_outcome.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'mutation_outcome.freezed.dart';
part 'mutation_outcome.g.dart';

@freezed
abstract class MutationOutcome with _$MutationOutcome {
  const factory MutationOutcome({
    @JsonKey(name: 'client_op_id') required String clientOpId,
    required MutationStatus status,
    @JsonKey(name: 'server_entity_id') String? serverEntityId,
    MutationError? error,
  }) = _MutationOutcome;

  factory MutationOutcome.fromJson(Map<String, Object?> json) =>
      _$MutationOutcomeFromJson(json);
}

enum MutationStatus {
  @JsonValue('APPLIED')
  applied,
  @JsonValue('REJECTED')
  rejected,
}

@freezed
abstract class MutationError with _$MutationError {
  const factory MutationError({
    required MutationErrorCode code,
    required String message,
  }) = _MutationError;

  factory MutationError.fromJson(Map<String, Object?> json) =>
      _$MutationErrorFromJson(json);
}

enum MutationErrorCode {
  @JsonValue('NOT_FOUND')
  notFound,
  @JsonValue('FORBIDDEN')
  forbidden,
  @JsonValue('INVALID_PAYLOAD')
  invalidPayload,
  @JsonValue('UNIQUE_VIOLATION')
  uniqueViolation,
  @JsonValue('CONFLICT')
  conflict,

  /// Granting OWNER role blocked because the target user already has a
  /// Member or Producer row.
  @JsonValue('OWNER_EXCLUSIVE')
  ownerExclusive,

  /// Granting PRODUCER role blocked because the target user already has a
  /// Member row or is OWNER.
  @JsonValue('PRODUCER_EXCLUSIVE')
  producerExclusive,

  /// Granting an AMAP role (ADMIN/COORDINATOR/VOLUNTEER) blocked because the
  /// target user is currently OWNER or PRODUCER.
  @JsonValue('MIXED_ROLES')
  mixedRoles,

  /// Removing ADMIN from the last admin of an organization is not allowed.
  @JsonValue('LAST_ADMIN')
  lastAdmin,

  /// Revoking the last OWNER of the instance is not allowed.
  @JsonValue('LAST_OWNER')
  lastOwner,

  /// Removing the last user with PRODUCER role from a producer organisation
  /// is not allowed.
  @JsonValue('LAST_PRODUCER')
  lastProducer,

  /// The caller targeted their own account for a destructive lifecycle
  /// action (suspend / reactivate / delete). Self-action is forbidden.
  @JsonValue('SELF_ACTION_FORBIDDEN')
  selfActionForbidden,

  /// A delivery cannot transition to CONFIRMED while at least one of its
  /// delivery contracts has no coordinator.
  @JsonValue('MISSING_COORDINATOR')
  missingCoordinator,

  /// A new member subscription or delivery link was rejected because the
  /// target contract's season has already ended (maxDeliveryDate < today).
  @JsonValue('CONTRACT_ENDED')
  contractEnded,

  /// A contract member subscription is empty or does not match any contract product price.
  @JsonValue('INVALID_SUBSCRIPTION')
  invalidSubscription,

  /// A contract shared basket is invalid: fewer than two members, a member that is not a
  /// contract member, a member appearing in two shared baskets, or members whose subscriptions
  /// are not identical.
  @JsonValue('INVALID_SHARED_BASKET')
  invalidSharedBasket,
}

File: back/persistence/wire/src/main/kotlin/MutationOutcome.kt
package persistence.changes

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Server-side outcome for a single [ClientMutation], correlated with the
 * client's mutation via [clientOpId].
 *
 * On [MutationStatus.APPLIED]:
 *  - The mutation has been committed atomically with its corresponding
 *    [Change] record. The client must consider the local mutation flushed.
 *  - For an UPSERT submitted with a temporary id ([ClientMutation.TMP_ID_PREFIX]),
 *    [serverEntityId] holds the real id allocated by the server. The client
 *    must rewrite its local row to use this id.
 *  - For all other cases, [serverEntityId] equals [ClientMutation.entityId].
 *
 * On [MutationStatus.REJECTED]:
 *  - Nothing was written. [error] explains why and the client should surface
 *    the failure to the user (or retry after fixing the input).
 */
@Serializable
data class MutationOutcome(
    @SerialName("client_op_id")
    val clientOpId: String,
    val status: MutationStatus,
    @SerialName("server_entity_id")
    val serverEntityId: String? = null,
    val error: MutationError? = null,
)

@Serializable
enum class MutationStatus {
    APPLIED,
    REJECTED,
}

@Serializable
data class MutationError(
    val code: MutationErrorCode,
    val message: String,
)

@Serializable
enum class MutationErrorCode {
    /** The targeted entity does not exist (or no longer exists). */
    NOT_FOUND,

    /** The mutation crosses a tenant boundary the caller does not own. */
    FORBIDDEN,

    /** The payload is missing, malformed, or fails type validation. */
    INVALID_PAYLOAD,

    /** Reserved: a domain-level uniqueness constraint was violated. */
    UNIQUE_VIOLATION,

    /** Reserved: optimistic-concurrency conflict (stale write). */
    CONFLICT,

    /**
     * Granting OWNER role blocked because the target user already has a Member or Producer row.
     * An AMAP-role user cannot also be OWNER.
     */
    OWNER_EXCLUSIVE,

    /**
     * Granting PRODUCER role blocked because the target user already has a Member row or is OWNER.
     * Symmetric to [OWNER_EXCLUSIVE].
     */
    PRODUCER_EXCLUSIVE,

    /**
     * Granting an AMAP role (ADMIN/COORDINATOR/VOLUNTEER) blocked because the target user is
     * currently OWNER or PRODUCER.
     */
    MIXED_ROLES,

    /** Removing ADMIN from the last admin of an organization is not allowed. */
    LAST_ADMIN,

    /** Revoking the last OWNER of the instance is not allowed. */
    LAST_OWNER,

    /**
     * Removing the last user with PRODUCER role from a producer organisation
     * is not allowed (the producer org would have nobody to represent it).
     */
    LAST_PRODUCER,

    /**
     * The caller targeted their own account for a destructive lifecycle
     * action (suspend / reactivate / delete). Self-action is forbidden to
     * avoid lock-out.
     */
    SELF_ACTION_FORBIDDEN,

    /**
     * A delivery in status CONFIRMED has at least one DeliveryContract whose
     * coordinators list is empty. The payload is rejected so the caller can
     * either assign a coordinator first, or move the delivery back to PLANNED.
     */
    MISSING_COORDINATOR,

    /**
     * A new member subscription or a new delivery link targets a contract whose
     * max_delivery_date is in the past. Existing entries remain editable.
     */
    CONTRACT_ENDED,

    /** A contract member subscription is empty or does not match any contract product price. */
    INVALID_SUBSCRIPTION,

    /**
     * A contract shared basket is invalid: fewer than two members, a member that is not a
     * contract member, a member appearing in two shared baskets, or members whose subscriptions
     * are not identical.
     */
    INVALID_SHARED_BASKET,
}
